### PR TITLE
[SPARK-5559] [Streaming] [Test] Remove oppotunity we met flakiness when running FlumeStreamSuite

### PR DESCRIPTION
**Pull request by sarutak**
When we run FlumeStreamSuite on Jenkins, sometimes we get error like as follows.

```
sbt.ForkMain$ForkError: The code passed to eventually never returned normally. Attempted 52 times over 10.094849836 seconds. Last failure message: Error connecting to localhost/127.0.0.1:23456.
    at org.scalatest.concurrent.Eventually$class.tryTryAgain$1(Eventually.scala:420)
    at org.scalatest.concurrent.Eventually$class.eventually(Eventually.scala:438)
    at org.scalatest.concurrent.Eventually$.eventually(Eventually.scala:478)
    at org.scalatest.concurrent.Eventually$class.eventually(Eventually.scala:307)
   at org.scalatest.concurrent.Eventually$.eventually(Eventually.scala:478)
   at org.apache.spark.streaming.flume.FlumeStreamSuite.writeAndVerify(FlumeStreamSuite.scala:116)
       at org.apache.spark.streaming.flume.FlumeStreamSuite.org$apache$spark$streaming$flume$FlumeStreamSuite$$testFlumeStream(FlumeStreamSuite.scala:74)
   at org.apache.spark.streaming.flume.FlumeStreamSuite$$anonfun$3.apply$mcV$sp(FlumeStreamSuite.scala:66)
    at org.apache.spark.streaming.flume.FlumeStreamSuite$$anonfun$3.apply(FlumeStreamSuite.scala:66)
    at org.apache.spark.streaming.flume.FlumeStreamSuite$$anonfun$3.apply(FlumeStreamSuite.scala:66)
    at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
    at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
    at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
    at org.scalatest.Transformer.apply(Transformer.scala:22)
    at org.scalatest.Transformer.apply(Transformer.scala:20)
        at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
    at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
    at org.scalatest.FunSuite.withFixture(FunSuite.scala:1555)
    at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
    at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
    at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
    at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
```

This error is caused by check-then-act logic  when it find free-port .

```
  /** Find a free port */
  private def findFreePort(): Int = {
    Utils.startServiceOnPort(23456, (trialPort: Int) => {
      val socket = new ServerSocket(trialPort)
      socket.close()
      (null, trialPort)
    }, conf)._2
  }
```

Removing the check-then-act is not easy but we can reduce the chance of having the error by choosing random value for initial port instead of 23456.

### CODE REVIEW

File: external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala [hunk 1]
```diff
@@ -76,7 +75,7 @@ class FlumeStreamSuite extends FunSuite with BeforeAndAfter with Matchers with L
 
   /** Find a free port */
   private def findFreePort(): Int = {
-    Utils.startServiceOnPort(23456, (trialPort: Int) => {
+    Utils.startServiceOnPort((math.random * Int.MaxValue).toInt, (trialPort: Int) => {
```
Comment (srowen) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala]: I like the idea but surely you don't mean to choose a random port from 0 to 2 billion? surely something in [1024,65536)? `1024 + new Random().nextInt(65536 - 1024)` or similar?

**Comment by JoshRosen** [comment]
I noticed that this `findFreePort` method is duplicated verbatim in MQQTSuite, so we should probably apply whatever fix we decide on there, too.

**Comment by sarutak** [comment]
@JoshRosen It's good if we could use the port `0` but I found there are no simple way to get the port where `FlumeReceiver#server (NettyServer)` bound.
I'm also investigating another flakiness in FlumeStreamSuite but I've not found so far...

And as you said, I found `findFreePort` in `MQTTStreamSuite` so I'll also fix it.

File: external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala (hunk 1 above)
Comment (sarutak) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala]: `Utils#startServiceOnPort` adjusts the range of port so I simply pass the random value to the method. But I noticed we should pass the value > 0.

      val tryPort = if (startPort == 0) {
        startPort
      } else {
        // If the new port wraps around, do not try a privilege port
        ((startPort + offset - 1024) % (65536 - 1024)) + 1024
      }

So, I'll fix the it.

Comment (srowen) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala]: It will actually fail if the very large random port, plus offset - 1024, rolls over to a negative value. The returned port will be negative. Although it's a rare case, I think writing the correct version is also clearer.

Comment (sarutak) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala]: Ah, I see. Should we modify `Utils#startServiceOnPort` ?

Comment (srowen) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala]: If anything, it should be changed to reject an invalid `startPort`, and this code should not pass invalid port values in any event.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -1814,6 +1814,12 @@ private[spark] object Utils extends Logging {
       startService: Int => (T, Int),
       conf: SparkConf,
       serviceName: String = ""): (T, Int) = {
+
+    if (startPort != 0 && (startPort < 1024 || startPort >= 65536)) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: I'd use `require()`; you can write `in [1024,65536)` to express the meaning succinctly; missing space after 0 and I don't think the last parens are needed.

File: external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala [hunk 3]
```diff
@@ -20,6 +20,8 @@ package org.apache.spark.streaming.flume
 import java.net.{InetSocketAddress, ServerSocket}
 import java.nio.ByteBuffer
 
+import org.apache.commons.lang3.RandomUtils
```
Comment (srowen) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumeStreamSuite.scala]: Doesn't this probably go after `scala.*` imports?

File: external/mqtt/src/test/scala/org/apache/spark/streaming/mqtt/MQTTStreamSuite.scala [hunk 4]
```diff
@@ -113,7 +115,8 @@ class MQTTStreamSuite extends FunSuite with Eventually with BeforeAndAfter {
   }
 
   private def findFreePort(): Int = {
```
Comment (srowen) [inline external/mqtt/src/test/scala/org/apache/spark/streaming/mqtt/MQTTStreamSuite.scala]: I wonder if this can be refactored rather than duplicated? don't bother unless there is clearly a good common home for it in test code. I don't know if there is.

Comment (JoshRosen) [inline external/mqtt/src/test/scala/org/apache/spark/streaming/mqtt/MQTTStreamSuite.scala]: We don't have a test utilities subproject, so this ends up getting duplicated, but note that we also have duplication of classes like LocalSparkContext; fixing this broader issue is outside the scope of this PR (there's a few JIRAs to track the creation of a "test utilities" project, though).

**Comment by andrewor14** [comment]
retest this please

**Comment by JoshRosen** [comment]
@sarutak If you've got time, mind addressing Sean's minor comments?  Other than those, this LGTM.

**Comment by sarutak** [comment]
Sorry, I had no time until last weekend but now I have. I'll address that soon.

**Comment by sarutak** [comment]
retest this please.

**Comment by srowen** [comment]
OK I think the import is still in slightly the wrong place in both cases and the text could be edited a bit, but I can carefully adjust that on merge.

**Comment by andrewor14** [comment]
Woohoo! Thanks @sarutak this has been flaky for a long time. Hopefully this fixes it.
